README
h1. HOTP - PHP Based HMAC One Time Passwords

CHANGELOG.md
# Changelog

# next (planned @ 2.0.1)
- Adopts PSR-12 syntax everywhere (:heart: [jrzepa](https://github.com/jrzepa))
- Expands version range for phpunit to ensure tests run against 7.2 [#16](https://github.com/jakobo/hotp-php/pull/16)

# 2.0.0 (Dec 20, 2020)
- Code coverage reporting added (:heart: [reedy](https://github.com/reedy))
- Documentation updates and return type hints (:heart: [reedy](https://github.com/reedy))
- Github CI Support [ac0d8d](https://github.com/jakobo/hotp-php/commit/ac0d8d0d64adc5f7ef83952bde25425bf74184cf) (:heart: [legoktm](https://github.com/legoktm))

## Breaking Changes
- **Bump minimum version** - To stay current, the minimum version was bumped from 5.3.3 to 7.2. We encourage you to update your PHP version to stay on top of potential security vulnerabilities and receive the latest performance and bug fixes.

# 1.0.1 (Apr 11, 2019)
- Token replay mitigation [d24e0d](https://github.com/jakobo/hotp-php/commit/d24e0d021710718cb9104ffb5c6ffb447fce65ab) (:heart: [reedy](https://github.com/reedy))
- Created `composer.json` and made available through composer (:heart: [reedy](https://github.com/reedy))

# 1.0.0
- [reedy](https://github.com/reedy) joined the maintainer team
- Initial Release with a tag

tests/HOTPTest.php
<?php

namespace jakobo\HOTP\Tests;

use jakobo\HOTP\HOTP;
use PHPUnit\Framework\TestCase;

/**
 * @copyright 2020
 * @license BSD-3-Clause
 * @version 1.0
 */
class HOTPTest extends TestCase
{
    private const KEY = '12345678901234567890';

    public function provideHOTP(): array
    {
        return [
            [
                0, [
                    'HMAC'  => 'cc93cf18508d94934c64b65d8ba7667fb7cde4b0',
                    'hex'   => '4c93cf18',
                    'dec'   => '1284755224',
                    'hotp'  => '755224',
                ]
            ],
            [
                1, [
                    'HMAC'  => '75a48a19d4cbe100644e8ac1397eea747a2d33ab',
                    'hex'   => '41397eea',
                    'dec'   => '1094287082',
                    'hotp'  => '287082',
                ]
            ],
            [
                2, [
                    'HMAC'  => '0bacb7fa082fef30782211938bc1c5e70416ff44',
                    'hex'   => '82fef30',
                    'dec'   => '137359152',
                    'hotp'  => '359152',
                ]
            ],
            [
                3, [
                    'HMAC'  => '66c28227d03a2d5529262ff016a1e6ef76557ece',
                    'hex'   => '66ef7655',
                    'dec'   => '1726969429',
                    'hotp'  => '969429',
                ]
            ],
            [
                4, [
                    'HMAC'  => 'a904c900a64b35909874b33e61c5938a8e15ed1c',
                    'hex'   => '61c5938a',
                    'dec'   => '1640338314',
                    'hotp'  => '338314',
                ]
            ],
            [
                5, [
                    'HMAC'  => 'a37e783d7b7233c083d4f62926c7a25f238d0316',
                    'hex'   => '33c083d4',
                    'dec'   => '868254676',
                    'hotp'  => '254676',
                ]
            ],
            [
                6, [
                    'HMAC'  => 'bc9cd28561042c83f219324d3c607256c03272ae',
                    'hex'   => '7256c032',
                    'dec'   => '1918287922',
                    'hotp'  => '287922',
                ]
            ],
            [
                7, [
                    'HMAC'  => 'a4fb960c0bc06e1eabb804e5b397cdc4b45596fa',
                    'hex'   => '4e5b397',
                    'dec'   => '82162583',
                    'hotp'  => '162583',
                ]
            ],
            [
                8, [
                    'HMAC'  => '1b3c89f65e6c9e883012052823443f048b4332db',
                    'hex'   => '2823443f',
                    'dec'   => '673399871',
                    'hotp'  => '399871',
                ]
            ],
            [
                9, [
                    'HMAC'  => '1637409809a679dc698207310c8c7fc07290d9e5',
                    'hex'   => '2679dc69',
                    'dec'   => '645520489',
                    'hotp'  => '520489',
                ]
            ],
        ];
    }

    /**
     * @param int $seed
     * @param array $result
     *
     * @covers \jakobo\HOTP\HOTP::generateByCounter
     * @covers \jakobo\HOTP\HOTPResult::toString
     * @covers \jakobo\HOTP\HOTPResult::toHex
     * @covers \jakobo\HOTP\HOTPResult::toDec
     * @covers \jakobo\HOTP\HOTPResult::toHOTP
     * @dataProvider provideHOTP
     */
    public function testHOTP(int $seed, array $result): void
    {
        $hotp = HOTP::generateByCounter(self::KEY, $seed);

        $this->assertEquals(
            $result['HMAC'],
            $hotp->toString()
        );

        $this->assertEquals(
            $result['hex'],
            $hotp->toHex()
        );

        $this->assertEquals(
            $result['dec'],
            $hotp->toDec()
        );

        $this->assertEquals(
            $result['hotp'],
            $hotp->toHOTP(6)
        );
    }

    public function provideTOTP(): array
    {
        return [
            [ '59', '94287082' ],
            [ '1111111109', '07081804' ],
            [ '1111111111', '14050471' ],
            [ '1234567890', '89005924' ],
            [ '2000000000','69279037'],
        ];
    }

    /**
     * @param string $seed
     * @param string $result
     *
     * @covers \jakobo\HOTP\HOTP::generateByTime
     * @covers \jakobo\HOTP\HOTPResult::toHOTP
     * @dataProvider provideTOTP
     */
    public function testTOTP(string $seed, string $result): void
    {
        $totp = HOTP::generateByTime(self::KEY, 30, $seed);

        $this->assertEquals(
            $result,
            $totp->toHOTP(8)
        );
    }

    public function provideGenerateByTimeWindow(): array
    {
        return [
                [ '1111111111', [
                    "404137",
                    "150727",
                    "731029",
                    "081804",
                    "050471",
                    "266759",
                    "306183",
                    "466594",
                    "754889"
                ]
            ]
        ];
    }

    /**
     * @param string $seed
     * @param string[] $result
     * @covers \jakobo\HOTP\HOTP::generateByTimeWindow
     * @covers \jakobo\HOTP\HOTPResult::toHOTP
     * @dataProvider provideGenerateByTimeWindow
     */
    public function testGenerateByTimeWindow(string $seed, array $result): void
    {
        $results = HOTP::generateByTimeWindow(
            self::KEY,
            30,
            -4,
            4,
            $seed
        );

        $actual = [];
        foreach ($results as $res) {
            $actual[] = $res->toHOTP(6);
        }

        $this->assertEquals(
            $result,
            $actual
        );
    }
}


hotp.php
<?php
/**
 * HOTP Class
 * Based on the work of OAuth, and the sample implementation of HMAC OTP
 * http://tools.ietf.org/html/draft-mraihi-oath-hmac-otp-04#appendix-D
 * @author Jakob Heuser (firstname)@felocity.com
 * @copyright 2011
 * @license BSD-3-Clause
 * @version 1.0
 */
class HOTP {
    /**
     * Generate a HOTP key based on a counter value (event based HOTP)
     * @param string $key the key to use for hashing
     * @param int $counter the number of attempts represented in this hashing
     * @return HOTPResult a HOTP Result which can be truncated or output
     */
    public static function generateByCounter($key, $counter) {
        // the counter value can be more than one byte long,
        // so we need to pack it down properly.
        $cur_counter = array(0, 0, 0, 0, 0, 0, 0, 0);
        for($i = 7; $i >= 0; $i--) {
            $cur_counter[$i] = pack ('C*', $counter);
            $counter = $counter >> 8;
        }
        
        $bin_counter = implode($cur_counter);
        
        // Pad to 8 chars
        if (strlen($bin_counter) < 8) {
            $bin_counter = str_repeat (chr(0), 8 - strlen ($bin_counter)) . $bin_counter;
        }

        // HMAC
        $hash = hash_hmac('sha1', $bin_counter, $key);
        
        return new HOTPResult($hash);
    }
    
    /**
     * Generate a HOTP key based on a timestamp and window size
     * @param string $key the key to use for hashing
     * @param int $window the size of the window a key is valid for in seconds
     * @param int $timestamp a timestamp to calculate for, defaults to time()
     * @return HOTPResult a HOTP Result which can be truncated or output
     */
    public static function generateByTime($key, $window, $timestamp = false) {
        if (!$timestamp && $timestamp !== 0) {
            $timestamp = HOTP::getTime();
        }
        
        $counter = intval($timestamp / $window);
        
        return HOTP::generateByCounter($key, $counter);
    }
    
    /**
     * Generate a HOTP key collection based on a timestamp and window size
     * all keys that could exist between a start and end time will be included
     * in the returned array
     * @param string $key the key to use for hashing
     * @param int $window the size of the window a key is valid for in seconds
     * @param int $min the minimum window to accept before $timestamp
     * @param int $max the maximum window to accept after $timestamp
     * @param int $timestamp a timestamp to calculate for, defaults to time()
     * @return array of HOTPResult
     */
    public static function generateByTimeWindow($key, $window, $min = -1, $max = 1, $timestamp = false) {
        if (!$timestamp && $timestamp !== 0) {
            $timestamp = HOTP::getTime();
        }
        
        $counter = intval($timestamp / $window);
        $window = range($min, $max);
        
        $out = array();
        for ($i = 0; $i < count($window); $i++) {
            $shift_counter = $window[$i];
            $out[$shift_counter] = HOTP::generateByCounter($key, $counter + $shift_counter);
        }
        
        return $out;
    }
    
    /**
     * Gets the current time
     * Ensures we are operating in UTC for the entire framework
     * Restores the timezone on exit.
     * @return int the current time
     */
    public static function getTime() {
        return time(); // PHP's time is always UTC
    }
}

/**
 * The HOTPResult Class converts an HOTP item to various forms
 * Supported formats include hex, decimal, string, and HOTP
 * @author Jakob Heuser (firstname)@felocity.com
 */
class HOTPResult {
    protected $hash;
    protected $binary;
    protected $decimal;
    
    /**
     * Build an HOTP Result
     * @param string $value the value to construct with
     */
    public function __construct($value) {
        // store raw
        $this->hash = $value;
    }
    
    /**
     * Returns the string version of the HOTP
     * @return string
     */
    public function toString() {
        return $this->hash;
    }
    
    /**
     * Returns the hex version of the HOTP
     * @return string
     */
    public function toHex() {
        if( !$this->hex ) 
        {
            $this->hex = dechex($this->toDec());
        }
        return $this->hex;
    }
    
    /**
     * Returns the decimal version of the HOTP
     * @return int
     */
    public function toDec() {
        if( !$this->decimal ) 
        {
            // store calculate decimal
            $hmac_result = array();
        
            // Convert to decimal
            foreach(str_split($this->hash,2) as $hex)
            {
               $hmac_result[] = hexdec($hex);
            }
        
            $offset = $hmac_result[19] & 0xf;
        
            $this->decimal = (
                (($hmac_result[$offset+0] & 0x7f) << 24 ) |
                (($hmac_result[$offset+1] & 0xff) << 16 ) |
                (($hmac_result[$offset+2] & 0xff) << 8 ) |
                ($hmac_result[$offset+3] & 0xff)
            );
        }
        return $this->decimal;
    }
    
    /**
     * Returns the truncated decimal form of the HOTP
     * @param int $length the length of the HOTP to return
     * @return string
     */
    public function toHOTP($length) {
        $str = str_pad($this->toDec(), $length, "0", STR_PAD_LEFT);
        $str = substr($str, (-1 * $length));
        return $str;
    }

}


src/HOTP.php
<?php

namespace jakobo\HOTP;

/**
 * HOTP Class
 * Based on the work of OAuth, and the sample implementation of HMAC OTP
 * http://tools.ietf.org/html/draft-mraihi-oath-hmac-otp-04#appendix-D
 * @author Jakob Heuser (firstname)@felocity.com
 * @copyright 2011-2020
 * @license BSD-3-Clause
 * @version 1.0
 */
class HOTP
{
    /**
     * Generate a HOTP key based on a counter value (event based HOTP)
     * @param string $key the key to use for hashing
     * @param int $counter the number of attempts represented in this hashing
     * @return HOTPResult a HOTP Result which can be truncated or output
     */
    public static function generateByCounter(string $key, int $counter): HOTPResult
    {
        // the counter value can be more than one byte long,
        // so we need to pack it down properly.
        $curCounter = [ 0, 0, 0, 0, 0, 0, 0, 0 ];
        for ($i = 7; $i >= 0; $i--) {
            $curCounter[$i] = pack('C*', $counter);
            $counter = $counter >> 8;
        }

        $binCounter = implode($curCounter);

        // Pad to 8 chars
        if (strlen($binCounter) < 8) {
            $binCounter = str_repeat(
                chr(0),
                8 - strlen($binCounter)
            ) . $binCounter;
        }

        // HMAC
        $hash = hash_hmac('sha1', $binCounter, $key);

        return new HOTPResult($hash);
    }

    /**
     * Generate a HOTP key based on a timestamp and window size
     * @param string $key the key to use for hashing
     * @param int $window the size of the window a key is valid for in seconds
     * @param int|false $timestamp a timestamp to calculate for, defaults to time()
     * @return HOTPResult a HOTP Result which can be truncated or output
     */
    public static function generateByTime(string $key, int $window, $timestamp = false): HOTPResult
    {
        if (!$timestamp && $timestamp !== 0) {
            // @codeCoverageIgnoreStart
            $timestamp = self::getTime();
            // @codeCoverageIgnoreEnd
        }

        $counter = intval($timestamp / $window) ;

        return self::generateByCounter($key, $counter);
    }

    /**
     * Generate a HOTP key collection based on a timestamp and window size
     * all keys that could exist between a start and end time will be included
     * in the returned array
     * @param string $key the key to use for hashing
     * @param int $window the size of the window a key is valid for in seconds
     * @param int $min the minimum window to accept before $timestamp
     * @param int $max the maximum window to accept after $timestamp
     * @param int|false $timestamp a timestamp to calculate for, defaults to time()
     * @return array of HOTPResult
     */
    public static function generateByTimeWindow(string $key, int $window, int $min = -1, int $max = 1, $timestamp = false): array
    {
        if (!$timestamp && $timestamp !== 0) {
            // @codeCoverageIgnoreStart
            $timestamp = self::getTime();
            // @codeCoverageIgnoreEnd
        }

        $counter = intval($timestamp / $window);
        $window = range($min, $max);

        $out = [];
        foreach ($window as $value) {
            $shiftCounter = $counter + $value;
            $out[$shiftCounter] = self::generateByCounter($key, $shiftCounter);
        }

        return $out;
    }

    /**
     * Gets the current time
     * Ensures we are operating in UTC for the entire framework
     * Restores the timezone on exit.
     * @return int the current time
     * @codeCoverageIgnore
     */
    public static function getTime(): int
    {
        // PHP's time is always UTC
        return time();
    }
}


src/HOTPResult.php
<?php

namespace jakobo\HOTP;

/**
 * The HOTPResult Class converts an HOTP item to various forms
 * Supported formats include hex, decimal, string, and HOTP

 * @author Jakob Heuser (firstname)@felocity.com
 * @copyright 2011-2020
 * @license BSD-3-Clause
 * @version 1.0
 */
class HOTPResult
{
    protected $hash;
    protected $decimal;
    protected $hex;

    /**
     * Build an HOTP Result
     * @param string $value the value to construct with
     * @codeCoverageIgnore
     */
    public function __construct(string $value)
    {
        // store raw
        $this->hash = $value;
    }

    /**
     * Returns the string version of the HOTP
     * @return string
     */
    public function toString(): string
    {
        return $this->hash;
    }

    /**
     * Returns the hex version of the HOTP
     * @return string
     */
    public function toHex(): string
    {
        if (!$this->hex) {
            $this->hex = dechex($this->toDec());
        }
        return $this->hex;
    }

    /**
     * Returns the decimal version of the HOTP
     * @return int
     */
    public function toDec(): int
    {
        if (!$this->decimal) {
            // store calculate decimal
            $hmacResult = [];

            // Convert to decimal
            foreach (str_split($this->hash, 2) as $hex) {
                $hmacResult[] = hexdec($hex);
            }

            $offset = $hmacResult[19] & 0xf;

            $this->decimal = (
                (($hmacResult[$offset+0] & 0x7f) << 24) |
                (($hmacResult[$offset+1] & 0xff) << 16) |
                (($hmacResult[$offset+2] & 0xff) << 8) |
                ($hmacResult[$offset+3] & 0xff)
            );
        }
        return $this->decimal;
    }

    /**
     * Returns the truncated decimal form of the HOTP
     * @param int $length the length of the HOTP to return
     * @return string
     */
    public function toHOTP(int $length): string
    {
        $str = str_pad($this->toDec(), $length, "0", STR_PAD_LEFT);
        return substr($str, (-1 * $length));
    }
}


example.php
<?php

/**
 * HOTP Example File
 */
require_once 'src/HOTP.php';

use jakobo\HOTP\HOTP;

$key = '12345678901234567890';

$table = [
    'HOTP' => [
        [
            'HMAC'  => 'cc93cf18508d94934c64b65d8ba7667fb7cde4b0',
            'hex'   => '4c93cf18',
            'dec'   => '1284755224',
            'hotp'  => '755224',
        ],
        [
            'HMAC'  => '75a48a19d4cbe100644e8ac1397eea747a2d33ab',
            'hex'   => '41397eea',
            'dec'   => '1094287082',
            'hotp'  => '287082',
        ],
        [
            'HMAC'  => '0bacb7fa082fef30782211938bc1c5e70416ff44',
            'hex'   => '82fef30',
            'dec'   => '137359152',
            'hotp'  => '359152',
        ],
        [
            'HMAC'  => '66c28227d03a2d5529262ff016a1e6ef76557ece',
            'hex'   => '66ef7655',
            'dec'   => '1726969429',
            'hotp'  => '969429',
        ],
        [
            'HMAC'  => 'a904c900a64b35909874b33e61c5938a8e15ed1c',
            'hex'   => '61c5938a',
            'dec'   => '1640338314',
            'hotp'  => '338314',
        ],
        [
            'HMAC'  => 'a37e783d7b7233c083d4f62926c7a25f238d0316',
            'hex'   => '33c083d4',
            'dec'   => '868254676',
            'hotp'  => '254676',
        ],
        [
            'HMAC'  => 'bc9cd28561042c83f219324d3c607256c03272ae',
            'hex'   => '7256c032',
            'dec'   => '1918287922',
            'hotp'  => '287922',
        ],
        [
            'HMAC'  => 'a4fb960c0bc06e1eabb804e5b397cdc4b45596fa',
            'hex'   => '4e5b397',
            'dec'   => '82162583',
            'hotp'  => '162583',
        ],
        [
            'HMAC'  => '1b3c89f65e6c9e883012052823443f048b4332db',
            'hex'   => '2823443f',
            'dec'   => '673399871',
            'hotp'  => '399871',
        ],
        [
            'HMAC'  => '1637409809a679dc698207310c8c7fc07290d9e5',
            'hex'   => '2679dc69',
            'dec'   => '645520489',
            'hotp'  => '520489',
        ],
    ],
    'TOTP' => [
        '59' => [
            'totp'  => '94287082',
        ],
        '1111111109' => [
            'totp'  => '07081804',
        ],
        '1111111111' => [
            'totp'  => '14050471',
        ],
        '1234567890' => [
            'totp'  => '89005924',
        ],
        '2000000000' => [
            'totp'  => '69279037',
        ],
    ],
];

echo <<<DOCBLOCK
<!DOCTYPE><html><head></head><body><pre>
http://www.ietf.org/rfc/rfc4226.txt
http://tools.ietf.org/html/draft-mraihi-totp-timebased-06

TEST VECTOR VERIFICATION

HOTP Tests:

DOCBLOCK;

echo "Count Method Value                                           Pass/Fail\n";
echo "----------------------------------------------------------------------\n";

// loop over all HOTP table results, and calculate the matching value
foreach ($table['HOTP'] as $seed => $results) {
    $hotp = HOTP::generateByCounter($key, $seed);
    $first = true;
    foreach ($results as $type => $calc) {
        if ($first) {
            echo str_pad($seed, 4, ' ', STR_PAD_LEFT);
            $first = false;
        } else {
            echo '    ';
        }
        echo '  ';
        echo str_pad($type, 5, ' ', STR_PAD_RIGHT);
        echo '  ';
        echo str_pad($calc, 47, ' ', STR_PAD_RIGHT);
        echo '  ';
        $method = 'to' . (ucfirst(str_replace('HMAC', 'string', $type)));
        echo str_pad(($calc == $hotp->$method(6)) ? '[OK]' : '[FAIL]', 9, ' ', STR_PAD_LEFT);
        echo "\n";
    }
}

echo <<<DOCBLOCK

TOTP Tests:

DOCBLOCK;

echo "Time (sec)   Value                                           Pass/Fail\n";
echo "----------------------------------------------------------------------\n";

// now echo over the TOTP table
foreach ($table['TOTP'] as $seed => $results) {
    $totp = HOTP::generateByTime($key, 30, $seed);
    $first = true;
    foreach ($results as $type => $calc) {
        if ($first) {
            echo str_pad($seed, 10, ' ', STR_PAD_LEFT);
            $first = false;
        } else {
            echo '          ';
        }
        echo '   ';
        echo str_pad($calc, 47, ' ', STR_PAD_RIGHT);
        echo '  ';
        $method = 'to' . (ucfirst(str_replace('totp', 'hotp', $type)));
        echo str_pad(($calc == $totp->$method(8)) ? '[OK]' : '[FAIL]', 9, ' ', STR_PAD_LEFT);
        echo "\n";
    }
}

echo '</pre></body></html>';
